File: packages/ref-sdk-py/alloist_ref_sdk/token.py
"""Token verification with JWKS (Ed25519)."""

from __future__ import annotations

from typing import Any

import jwt
from jwt import PyJWKSet
from jwt.exceptions import ExpiredSignatureError, InvalidTokenError


def verify_token(token: str, jwks: dict[str, Any]) -> dict[str, Any]:
    """
    Verify ACT-lite JWT with JWKS.
    Returns {valid: True, jti, scopes, subject} or {valid: False, reason}.
    """
    try:
        jwk_set = PyJWKSet.from_dict(jwks)
        header = jwt.get_unverified_header(token)
        kid = header.get("kid")
        signing_key = None
        for key in jwk_set.keys:
            if key.key_id == kid:
                signing_key = key
                break
        if not signing_key:
            return {"valid": False, "reason": "key_not_found"}

        payload = jwt.decode(
            token,
            signing_key.key,
            algorithms=["EdDSA"],
            options={"verify_exp": True},
            leeway=5,
        )
        return {
            "valid": True,
            "jti": payload.get("jti", ""),
            "scopes": payload.get("scopes", []),
            "subject": payload.get("sub", ""),
        }
    except ExpiredSignatureError:
        return {"valid": False, "reason": "token_expired"}
    except jwt.InvalidSignatureError:
        return {"valid": False, "reason": "invalid_signature"}
    except InvalidTokenError as e:
        return {"valid": False, "reason": str(e) if str(e) else "invalid_token"}

File: packages/ref-sdk-node/src/token.js
'use strict';

const { createLocalJWKSet, jwtVerify } = require('jose');

/**
 * Verify ACT-lite JWT with JWKS.
 * @param {string} token - JWT string
 * @param {object} jwks - JWKS object { keys: [...] }
 * @returns {{ valid: true, jti: string, scopes: string[], subject: string } | { valid: false, reason: string }}
 */
async function verifyToken(token, jwks) {
  const JWKS = createLocalJWKSet(jwks);
  try {
    const { payload } = await jwtVerify(token, JWKS, {
      algorithms: ['EdDSA'],
      clockTolerance: 5,
    });
    const exp = payload.exp;
    if (exp && exp < Math.floor(Date.now() / 1000)) {
      return { valid: false, reason: 'token_expired' };
    }
    return {
      valid: true,
      jti: payload.jti || '',
      scopes: payload.scopes || [],
      subject: payload.sub || '',
    };
  } catch (err) {
    if (err.code === 'ERR_JWT_EXPIRED') {
      return { valid: false, reason: 'token_expired' };
    }
    if (err.code === 'ERR_JWS_SIGNATURE_VERIFICATION_FAILED') {
      return { valid: false, reason: 'invalid_signature' };
    }
    if (err.code === 'ERR_JWKS_NO_MATCHING_KEY' || err.message?.includes('kid')) {
      return { valid: false, reason: 'key_not_found' };
    }
    return { valid: false, reason: err.message || 'invalid_token' };
  }
}

module.exports = { verifyToken };
